src/sources/exec.rs

```diff
@@ -0,0 +1,868 @@
+
+                spawn_streaming_thread(stderr_lines, stderr_shutdown, STDERR, error_sender);
+            }
+
+            spawn_streaming_thread(stdout_lines, stdout_shutdown, STDOUT, sender);
+
+            while let Some((line, stream)) = receiver.recv().await {
+                let event = create_event(
+                    &config,
+                    Bytes::from(line),
+                    &None,
+                    &Some(stream.to_string()),
+                    Some(pid),
+                    None,
+                    &None,
+                );
+
+                out.send(event)
+                    .await
+                    .map_err(|_: crate::pipeline::ClosedError| {
+                        error!(message = "Failed to forward events; downstream is closed.");
+                    })?;
+            }
+
+            info!("Finished streaming exec run.");
+            let _ = out.flush().await;
+            Ok(())
+        }
+        Err(spawn_error) => {
+            error!("Error during streaming exec run.");
+            emit!(ExecFailed { error: spawn_error });
+
+            Err(())
+        }
+    }
+}
+
+async fn run_once_scheduled(
+    config: &ExecConfig,
+    exec_interval_secs: u64,
+    event_per_line: bool,
+    exec_duration_millis_key: String,
+) -> Option<Vec<Event>> {
+    let mut command = build_command(config);
+
+    // Mark the start time just before spawning the process as
+    // this seems to be the best approximation of exec duration
+    let now = Instant::now();
+    let child = command.spawn();
+
+    match child {
+        Ok(child) => {
+            let schedule = Duration::from_secs(exec_interval_secs);
+            let pid = child.id();
+
+            // Wait for our task to finish, wrapping it in a timeout
+            let timeout = tokio::time::timeout(schedule, child.wait_with_output());
+            let timeout_result = timeout.await;
+
+            match timeout_result {
+                Ok(output) => match output {
+                    Ok(output) => {
+                        let mut events = Vec::new();
+                        let elapsed_millis = now.elapsed().as_millis();
+                        let exit_status = output.status.code();
+
+                        let stdout_events = process_scheduled_output(
+                            config,
+                            event_per_line,
+                            exec_duration_millis_key.clone(),
+                            pid,
+                            exit_status,
+                            elapsed_millis,
+                            output.stdout,
+                            STDOUT,
+                        );
+
+                        if let Some(mut stdout_events) = stdout_events {
+                            events.append(&mut stdout_events);
+                        }
+
+                        // Optionally include stderr
+                        if config.include_stderr.unwrap_or(true) {
```

Comment:
What happens to `child.stderr` if nothing is reading from it (i.e. `include_stderr = false`? I'm wondering if it'll block.

Reply:
I have tested running a command that produced output on stdout and stderr. When I don't collect stderr everything still shutdown nicely.